### PR TITLE
ALPN implementation for the updated SSLServiceDelegate

### PR DESCRIPTION
**Pull request by ymesika**
Application-Layer Protocol Negotiation (ALPN) is a TLS extension allows the application-layer to negotiate which protocol will by used over the secured socket.
The main motivation for adding this support to BlueSSLService is to add support for HTTP/2 which uses this extension to choose the newer version of HTTP over the older ones. 

## Description
This PR follows the [BlueSocket PR #67](https://github.com/IBM-Swift/BlueSocket/pull/67) that added new field/function to the SSLServiceDelegate protocol.

In this PR the implementation uses the OpenSSL API to register callback functions that will negotiate for a protocol based on ALPN information arriving from the client within the ClientHello message.
The implementation tries to match a supported protocol from the **availableAlpnProtocols** list with the protocols requested by the client. If it will find a match the response will be sent back to the client and the selected protocol will be set in the **negotiatedAlpnProtocol** field.

Because of limitations of the Apple Secure Transport it is still unclear how and if ALPN negotiation is possible on MacOS.
In this PR we only support the Linux environment by integrating with the OpenSSL API.

## Motivation and Context
The main motivation for adding this support to BlueSSLService is to add support for HTTP/2 which uses this extension to choose the newer version of HTTP over the older ones.

## How Has This Been Tested?
It has been tested with broader changes to Kitura-net and a new package under development for adding HTTP/2 support to Kitura.

## Checklist:

- [ ] I have submitted a [CLA form](https://github.com/IBM-Swift/CLA)
- [ ] If applicable, I have updated the documentation accordingly.
- [ ] If applicable, I have added tests to cover my changes.

### CODE REVIEW

**Comment by ymesika** [comment]
Closing the PR until I find a way to check the OpenSSL version and make sure it's >= 1.0.2.
Currently it fails the testings.

**Comment by ymesika** [comment]
@shmuelk FYI

**Comment by billabt** [comment]
Please note also that the macOS build failed as well.

**Comment by ymesika** [comment]
Thanks @billabt I fixed the macOS issues.
However, I still need to figure out how to do something similar to `#if OPENSSL_VERSION_NUMBER >= 0x10002000L` in Swift.
Therefore not reopening yet.

**Comment by ymesika** [comment]
I'm now testing the OpenSSL version number with the SSLeay() which can be found in any version of the library.
Reopening to make sure all build errors are now fixed.

**Comment by ymesika** [comment]
Logic for checking the OpenSSL version will be moved to the OpenSSL package.
Will be reopened once the OpenSSL PR (TBD) will be merged.

**Comment by ymesika** [comment]
Reopening after the OpenSSL PR #3 dependency has been merged making this PR ready to go.

File: Sources/SSLService.swift [hunk 1]
```diff
@@ -330,6 +333,20 @@ public class SSLService: SSLServiceDelegate {
 		public private(set) var context: SSLContext?
 	
 	#endif
+
+	// MARK: ALPN
```
Comment (billabt) [inline Sources/SSLService.swift]: The ALPN protocol is currently only for Linux.  Please wrapper the implementation (where appropriate) in _#if os(Linux)_.

Comment (billabt) [inline Sources/SSLService.swift]: Please update code to the latest version in _master_.  I merged another PR today to correct a crash problem that changed a few things.  Thanks.

**Pull request merged**